core/src/main/java/org/elasticsearch/index/engine/KeepUntilGlobalCheckpointDeletionPolicy.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.index.IndexCommit;
+import org.elasticsearch.common.lucene.index.ESIndexDeletionPolicy;
+import org.elasticsearch.index.seqno.SequenceNumbers;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.LongSupplier;
+
+/**
+ * An {@link ESIndexDeletionPolicy} that deletes index commits that are not required for recovery.
+ * In particular, this policy will delete index commits whose max sequence number is smaller (or equal) than
+ * the current global checkpoint except the index commit which has the highest max sequence number among those.
+ */
+final class KeepUntilGlobalCheckpointDeletionPolicy implements ESIndexDeletionPolicy {
+    private final LongSupplier globalCheckpointSupplier;
+
+    KeepUntilGlobalCheckpointDeletionPolicy(LongSupplier globalCheckpointSupplier) {
+        this.globalCheckpointSupplier = globalCheckpointSupplier;
+    }
+
+    @Override
+    public List<IndexCommit> onInit(List<? extends IndexCommit> commits) throws IOException {
+        if (commits.isEmpty()) {
+            return Collections.emptyList();
+        }
+        return onCommit(commits);
+    }
+
+    @Override
+    public List<IndexCommit> onCommit(List<? extends IndexCommit> commits) throws IOException {
+        assert commits.isEmpty() == false : "onCommit() must be called with a non-empty list of commits";
+
+        final List<IndexCommit> keptCommits = new ArrayList<>();
+        final int keptPosition = indexOfKeptCommits(commits);
+        final List<Integer> duplicateIndexes = indexesOfDuplicateCommits(commits);
```

Comment:
We only call this collection's `contains()` method so maybe a `HashSet<Integer>` would be better? Not sure if it gets big enough to make a difference.

Reply:
We expect to have zero or one element in a list, there should be no difference. However, using Set makes more sense here; I updated. Thank you.